R/position-.R

```diff
@@ -66,6 +68,36 @@ Position <- ggproto("Position",
 
   compute_panel = function(self, data, params, scales) {
     cli::cli_abort("Not implemented.")
+  },
+
+  aesthetics = function(self) {
+    required_aes <- self$required_aes
+    if (!is.null(required_aes)) {
+      required_aes <- unlist(strsplit(self$required_aes, "|", fixed = TRUE))
+    }
+    c(union(required_aes, names(self$default_aes)))
+  },
+
+  use_defaults = function(self, data, params = list()) {
+
+    aes <- self$aesthetics()
+    defaults <- self$default_aes
+
+    params <- params[intersect(names(params), aes)]
+    params <- params[setdiff(names(params), names(data))]
+    defaults <- defaults[setdiff(names(defaults), c(names(params), names(data)))]
+
+    if ((length(params) + length(defaults)) < 1) {
+      return(data)
+    }
+
+    new <- compact(lapply(defaults, eval_tidy, data = data))
+    new[names(params)] <- params
+    check_aesthetics(new, nrow(data))
+
+    data[names(new)] <- new
+    data
+
```

Comment:
There are some restrictions here that don't apply to the classic aesthetics, right? You can't use `after_scale()` for instance. Is that somehow taken care of?

Reply:
Any surviving modifiers would be ignored. `after_stat()` is already evaluated, so that isn't really a concern. `after_scale()` doesn't really apply as position adjustments are computed before scales are applied. It will be interpreted as there is no such aesthetic, therefore the default value is substituted.

I'm not sure whether this was clear, but this does not replace `Geom$use_defaults()`, it just evaluates the `Position$default_aes` really.